libs/pdi-scanner/src/scanner.rs

```diff
@@ -0,0 +1,265 @@
+
+        Ok(Self {
+            device_handle,
+            default_timeout: Duration::from_secs(1),
+            stop_tx: None,
+            thread_handle: None,
+        })
+    }
+
+    pub fn set_default_timeout(&mut self, timeout: Duration) {
+        self.default_timeout = timeout;
+    }
+
+    pub fn start(
+        &mut self,
+    ) -> (
+        Sender<(usize, packets::Outgoing)>,
+        Receiver<usize>,
+        Receiver<packets::Incoming>,
+    ) {
+        /// The endpoint for sending commands to the scanner.
+        const ENDPOINT_OUT: u8 = 0x05;
+
+        /// The primary endpoint for receiving responses from the scanner.
+        const ENDPOINT_IN_PRIMARY: u8 = 0x85;
+
+        /// The alternate endpoint for receiving responses from the scanner, used to
+        /// receive image data.
+        const ENDPOINT_IN_IMAGE_DATA: u8 = 0x86;
+
+        const BUFFER_SIZE: usize = 16_384;
+
+        let (host_to_scanner_tx, host_to_scanner_rx) =
+            mpsc::channel::<(usize, packets::Outgoing)>();
+        let (host_to_scanner_ack_tx, host_to_scanner_ack_rx) = mpsc::channel::<usize>();
+        let (scanner_to_host_tx, scanner_to_host_rx) = mpsc::channel();
+        let (stop_tx, stop_rx) = mpsc::channel();
+
+        self.stop_tx = Some(stop_tx);
+
+        let mut transfer_pool = rusb_async::TransferPool::new(self.device_handle.clone());
+        let default_timeout = self.default_timeout;
+
+        self.thread_handle = Some(thread::spawn({
+            move || {
+                tracing::debug!("Scanner thread started; submitting initial IN endpoint transfers");
+                transfer_pool
+                    .submit_bulk(ENDPOINT_IN_PRIMARY, vec![0; BUFFER_SIZE])
+                    .unwrap();
+
+                transfer_pool
+                    .submit_bulk(ENDPOINT_IN_IMAGE_DATA, vec![0; BUFFER_SIZE])
+                    .unwrap();
+
+                let mut is_stopping = false;
+                loop {
+                    match stop_rx.try_recv() {
+                        Ok(StopMode::CancelPendingTransfers) => {
+                            tracing::debug!(
+                                "Scanner thread received stop signal; cancelling all transfers"
+                            );
+                            transfer_pool.cancel_all();
+                            break;
+                        }
+                        Ok(StopMode::WaitUntilTransfersComplete) => {
+                            tracing::debug!(
+                                "Scanner thread received stop signal; cancelling incoming transfers and waiting for outgoing transfers to complete"
+                            );
+                            is_stopping = true;
+                            transfer_pool.cancel_all_for_endpoint(ENDPOINT_IN_PRIMARY);
+                            transfer_pool.cancel_all_for_endpoint(ENDPOINT_IN_IMAGE_DATA);
+                        }
+                        Err(_) => {}
+                    }
```

Comment:
Why have two stop modes? It's not totally clear to me how to choose one over the other

Reply:
We could make a call that we always use one. In your debugging I think you found that canceling the pending transfers actually caused issues, right? So maybe we just don't.

Reply:
Changed this to not have `StopMode`.